### PR TITLE
Feature/kustomize implementation

### PR DESCRIPTION
**Pull request by kelsoncarmozinezup**
## Description
Add Kustomization support and integrations tests for it.

### CODE REVIEW

File: internal/client/kustomization.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+type Kustomization struct {
+	manifests ManifestsReader
+}
+
+func NewKustomization(manifests ManifestsReader) *Kustomization {
+	return &Kustomization{manifests: manifests}
+}
+
+func (k *Kustomization) Template(ctx context.Context, source, path string) (mf.Manifest, error) {
```
Comment (thallesfreitaszup) [inline internal/client/kustomization.go:40]: Isn't Kustomize a more semantic name than Template? I think  template is more related to helm than kustomize.

File: pkg/module/kustomization_validation.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+)
+
+func NewKustomizationValidation(kustomization KustomizationClient, status StatusWriter) *KustomizationValidation {
+	return &KustomizationValidation{kustomization: kustomization, status: status}
+}
+
+func (k *KustomizationValidation) Reconcile(ctx context.Context, obj client.Object) (ctrl.Result, error) {
+	module, ok := obj.(*charlescdv1alpha1.Module)
+	log := logr.FromContextOrDiscard(ctx)
+	log.Info("kustomizationValid")
```
Comment (thallesfreitaszup) [inline pkg/module/kustomization_validation.go:55]: same here

File: internal/client/kustomization.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	r, err := json.Marshal(resMap.Resources())
+	if err != nil {
+		return mf.Manifest{}, fmt.Errorf("erro on resMap marshal: %w", err)
+	}
+
+	return k.manifests.FromBytes(ctx, r)
+}
+
+func (k *Kustomization) template(filePath string) (resmap.ResMap, error) {
```
Comment (tiagoangelozup) [inline internal/client/kustomization.go:64]: I tried to create a module with the below resource:
```yaml
apiVersion: deploy.charlescd.io/v1alpha1
kind: Module
metadata:
  name: football-bets
kustomization:
  gitRepository:
    path: deploy
    ref:
      type: commit
      value: bcf5f8101a1ef433013152ac89b3502b58234365
    url: https://github.com/tiagoangelozup/football-bets
```
but I got the following error
```yaml
status:
  conditions:
    - lastTransitionTime: "2022-04-01T12:30:48Z"
      message: Sources available locally at /tmp/gitrepository/kuttl-test-worthy-escargot/football-bets/bcf5f8101a1ef433013152ac89b3502b58234365.tar.gz
      reason: Downloaded
      status: "True"
      type: SourceReady
    - lastTransitionTime: "2022-04-01T12:30:48Z"
      message: 'erro running kustomizer: %!w(<nil>)'
      reason: RenderError
      status: "False"
      type: SourceValid
  phase: Failed
  source:
    path: /tmp/gitrepository/kuttl-test-worthy-escargot/football-bets/bcf5f8101a1ef433013152ac89b3502b58234365.tar.gz
```
here are some logs [charlescd.log](https://github.com/ZupIT/charlescd-operator/files/8397862/charlescd.log)

File: internal/client/kustomization.go (hunk 1 above)
Comment (kelsoncarmozinezup) [reply internal/client/kustomization.go]: Changed to kustomize.

File: pkg/module/helm_validation.go [hunk 4]
```diff
@@ -51,6 +51,8 @@ func NewHelmValidation(helm HelmClient, status StatusWriter) *HelmValidation {
 
 func (h *HelmValidation) Reconcile(ctx context.Context, obj client.Object) (ctrl.Result, error) {
 	module, ok := obj.(*charlescdv1alpha1.Module)
+	log := logr.FromContextOrDiscard(ctx)
+	log.Info("HelmValid")
```
[the remark this answers is not in the document]
Comment (kelsoncarmozinezup) [reply pkg/module/helm_validation.go]: Log removed.

File: pkg/module/kustomization_validation.go (hunk 2 above)
Comment (kelsoncarmozinezup) [reply pkg/module/kustomization_validation.go]: Log removed.

File: go.mod [hunk 5]
```diff
@@ -21,11 +21,13 @@ require (
...
-	k8s.io/api v0.23.4
+	k8s.io/api v0.23.5
 	k8s.io/apiextensions-apiserver v0.23.4
-	k8s.io/apimachinery v0.23.4
-	k8s.io/client-go v0.23.4
+	k8s.io/apimachinery v0.23.5
+	k8s.io/client-go v0.23.5
 	sigs.k8s.io/controller-runtime v0.11.1
+	sigs.k8s.io/kustomize/api v0.11.2
+	sigs.k8s.io/kustomize/kustomize/v4 v4.4.1
```
[the remark this answers is not in the document]
Comment (kelsoncarmozinezup) [reply go.mod]: Updated libraries.

File: internal/client/kustomization.go (hunk 3 above)
Comment (kelsoncarmozinezup) [reply internal/client/kustomization.go]: Bug fixed.

**Review by thallesfreitaszup — DISMISSED** [review]

File: go.mod [hunk 6]
```diff
@@ -21,11 +21,14 @@ require (
 	go.opentelemetry.io/otel/trace v1.4.1
 	go.uber.org/zap v1.21.0
 	helm.sh/helm/v3 v3.8.0
-	k8s.io/api v0.23.4
+	k8s.io/api v0.23.5
 	k8s.io/apiextensions-apiserver v0.23.4
```
Comment (tiagoangelozup) [inline go.mod:25]: usually `k8s.io/apiextensions-apiserver` match the same versions of `k8s.io/apimachinery` and `k8s.io/client-go`.

File: pkg/module/kustomization_validation.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	charlescdv1alpha1 "github.com/ZupIT/charlescd-operator/api/v1alpha1"
+	"github.com/ZupIT/charlescd-operator/internal/tracing"
+)
+
+const (
+	sourceError              = "Error templating source"
+	kustomizationValid       = "Kustomize configurations is valid"
```
Comment (tiagoangelozup) [inline pkg/module/kustomization_validation.go:32]: "Kustomize configurations **are** valid"

**Review by barbararochazup — CHANGES_REQUESTED** [review]
Add a pull request description

**Review by tiagoangelozup — APPROVED** [review]

**Review by barbararochazup — DISMISSED** [review]

**Review by tiagoangelozup — DISMISSED** [review]

**Review by thallesfreitaszup — APPROVED** [review]

**Review by tiagoangelozup — APPROVED** [review]

**Review by barbararochazup — APPROVED** [review]

**Pull request merged**